### PR TITLE
Add IN_PARTITIONED_SUBQUERY support

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Add `IN_PARTITIONED_SUBQUERY` transform function to support `IDSET` aggregation function as the subquery on the server side. Because the subquery is solved on the server side, in order to make it work, the subquery must hit the same table as the main query, and the table must be partitioned at server level (all the segments for a partition is served by a single server).

E.g. The following 2 queries can be combined into one query:
`SELECT ID_SET(col) FROM table WHERE date = 20200901`
`SELECT DISTINCT_COUNT(col), date FROM table WHERE IN_ID_SET(col, '<serializedIdSet>') = 1 GROUP BY date`
->
`SELECT DISTINCT_COUNT(col), date FROM table WHERE IN_PARTITIONED_SUBQUERY(col, 'SELECT ID_SET(col) FROM table WHERE date = 20200901') = 1 GROUP BY date`

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 1]
```diff
@@ -277,4 +253,112 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
...
+        handleSubquery(child, indexSegments, timerContext, executorService, endTimeMs);
+      }
+    }
+  }
+
+  /**
+   * Handles the subquery in the given expression.
+   * <p>Currently only supports subquery within IN_PARTITIONED_SUBQUERY function.
+   */
+  private void handleSubquery(ExpressionContext expression, List<IndexSegment> indexSegments, TimerContext timerContext,
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:337]: add java docs on what this method is doing

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 2]
```diff
@@ -277,4 +253,112 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
...
+
+  /**
+   * Handles the subquery in the filter.
+   * <p>Currently only supports subquery within IN_PARTITIONED_SUBQUERY function.
+   */
+  private void handleSubquery(FilterContext filter, List<IndexSegment> indexSegments, TimerContext timerContext,
+      ExecutorService executorService, long endTimeMs)
+      throws Exception {
+    Predicate predicate = filter.getPredicate();
+    if (predicate != null) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:324]: docs, isn't it better to check for filter.getChildren and invert the if statement?

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 3]
```diff
@@ -277,4 +253,112 @@ public DataTable processQuery(ServerQueryRequest queryRequest, ExecutorService e
...
+    }
+    List<ExpressionContext> arguments = function.getArguments();
+    if (StringUtils.remove(function.getFunctionName(), '_').equalsIgnoreCase(IN_PARTITIONED_SUBQUERY)) {
+      Preconditions
+          .checkState(arguments.size() == 2, "IN_PARTITIONED_SUBQUERY requires 2 arguments: expression, subquery");
+      ExpressionContext subqueryExpression = arguments.get(1);
+      Preconditions.checkState(subqueryExpression.getType() == ExpressionContext.Type.LITERAL,
+          "Second argument of IN_PARTITIONED_SUBQUERY must be a literal (subquery)");
+      QueryContext subquery = QueryContextConverterUtils.getQueryContextFromSQL(subqueryExpression.getLiteral());
+      DataTable dataTable =
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:352]: more javadocs

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java [hunk 4]
```diff
@@ -372,5 +399,46 @@ private static void getAggregations(FilterContext filter, List<FunctionContext>
         getAggregations(filter.getPredicate().getLhs(), aggregations);
       }
     }
+
+    /**
+     * Helper method to extract the columns (IDENTIFIER expressions) for the query.
+     */
+    public void extractColumns(QueryContext query) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:406]: why is this a public method? who is invoking this and from where

**Comment by yupeng9** [comment]
High-level comments on the signature of `IN_PARTITIONED_SUBQUERY(col, 'SELECT ID_SET(col) FROM table WHERE date = 20200901')`

Instead of taking a subquery as the last argument, shall we consider an expression? With the function `IN_PARTITIONED_SUBQUERY` and the first argument `col`, we can already construct the subquery together with the expression for the where clause.

**Comment by kishoreg** [comment]
> High-level comments on the signature of `IN_PARTITIONED_SUBQUERY(col, 'SELECT ID_SET(col) FROM table WHERE date = 20200901')`
> 
> Instead of taking a subquery as the last argument, shall we consider an expression? With the function `IN_PARTITIONED_SUBQUERY` and the first argument `col`, we can already construct the subquery together with the expression for the where clause.

Can you please elaborate - maybe an example will help

**Comment by yupeng9** [comment]
Looking at the implementation, the subquery can take only predicates and which is translated to a filter. Then we might not want `IN_PARTITIONED_SUBQUERY ` to take a free-formed SQL query as the argument.

The signature can be : 
```
IN_PARTITIONED_SUBQUERY (col, col_filtrerTable, filterTable, conditions_filterTable)
```

**Comment by siddharthteotia** [comment]
I think we are having mixed semantics here. Even though it is a combination of 2 queries, are implementation is a transform function. It is not a correct subquery syntax that we usually see in other systems. 

The name of the transform functions should not contain subquery since it is implementation detail. On that  note, don't think transform function should even be used to implement subquery semantics in Pinot since that is not the standard SQL semantics. 

We should just have the inner query in parentheses to implement a proper non correlated subquery in SQL. Is it absolutely necessary to have this subquery specified in a transform function?

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:337]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:324]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:352]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/QueryContext.java:406]: Good catch, changed to private

**Comment by Jackie-Jiang** [comment]
> Looking at the implementation, the subquery can take only predicate which is translated to a filter. Then we might not want `IN_PARTITIONED_SUBQUERY ` to take a free-formed SQL query as the argument.
> 
> The signature can be :
> 
> ```
> IN_PARTITIONED_SUBQUERY (col, col_filtrerTable, filterTable, conditions_filterTable)
> ```

@yupeng9 Added some validation on the subquery. I prefer having the sql query as the argument because that is more flexible, and we can add other subquery support easier in the future.

**Comment by Jackie-Jiang** [comment]
> I think we are having mixed semantics here. Even though it is a combination of 2 queries, our implementation is a transform function. It is not a correct subquery syntax that we usually see in other systems.
> 
> The name of the transform functions should not contain subquery since it is implementation detail. On that note, don't think transform function should even be used to implement subquery semantics in Pinot since that is not the standard SQL semantics.
> 
> We should just have the inner query in parentheses to implement a proper non correlated subquery in SQL. Is it absolutely necessary to have this subquery specified in a transform function?

@siddharthteotia Currently our query parser & PinotQuery does not support inner query, thus we use transform function as a work around for this. In the future when we have the inner query support, we can change this to use inner query if necessary. This feature is in experiment phase, and I don't want to block this on the nested query project which has much bigger scope.

**Review by kishoreg — APPROVED** [review]

**Pull request merged**